Integrate C++ into the PSyKAl model (manual version)

sergisiso opened the pull request:
**Fortran** PSyKAl implementations look like this:

Algorithm -> PSy-layer -> Kernels
+Intrastructure(Alg: calls and globals, PSy-layer: calls when DM and globals, Kernels: globals)

In the case of nemolite2d manual implementations this are:
Algorithm: nemolite2d.f90
Psy-layer: time_step_mod.f90
Kernels: multiple .f90 files
Infrastructure: dl_esm_inf


To integrate **C/OpenCl** we used the following aproach:
 - Kernels had to be transformed to the target language
 - Infrastructure globals transformed to arguments (a PSyclone transformation can do this)
 - PSy-layer call fortran iso_c_bindings wrappers to c host calls to set up OpenCL and execute its kernels:

Algorithm -> PSy-layer -> OpenCL Lib wrapper-> Kernels

Algorithm: nemolite2d.f90
Psy-layer: time_step_mod.f90
Kernels: multiple .cl files
Wrapper: FortCL
Infrastructure: dl_esm_inf

The interoperability between Fortran and **C++** is limited (e.g. templates not supported). Some of this are essential for efficient kokkos and SYCL use. To integrate C++ in existing Fortran psyclone applications ideally we want to keep Algorithm and Infrastructure as they are (Fortran in nemolite2d).

There are 2 alternatives:
	- Again do a wrapper or utility library (for each target?)
	- Do the whole PSy-layer with the target language and the call from Algorithm->PSy-layer with iso_c_binding to an `extern "C"` function.

sergisiso commented:
@rupertford @arporter As I explained in the telco I decided to go for the 2 layer option: The first in Fortran to fetch the infrastructure data and then a iso_c_binding interface to the second layer in the target (c family) language.

The relevant files are:
- [Fortran layer](https://github.com/stfc/PSycloneBench/blob/cpp_baseline/benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step_mod.f90)
- [C++ PSylayer](https://github.com/stfc/PSycloneBench/blob/cpp_baseline/benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step.cpp)

I think this should allow for extending  a kokkos/sycl/modern cpp version by just modifying the `time_step.cpp` file. And I think is feasible that PSyclone eventually generates something like this.

Performance-wise is in the same ballpark than Fortran, the table below is 1024x1024x100it. (this is when both versions manage to inline the kernels)
```
Implementation          Compiler        Arch    ua checksum     uv checksum     time/step
psykal_serial:          gcc             cpu     0.52670513E+02  0.57100321E+03  0.18978E+00
psykal_cpp:             gcc             cpu     0.52670513E+02  0.57100321E+03  0.15740E+00
```

I will stop this PR here and continue other C++ implementations in follow-up PRs. So this is ready for review.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/Makefile:
```diff
@@ -0,0 +1,88 @@
+# This Makefile picks up the compiler to use any flags from
+# environment variables. e.g. to use gfortran:
+# 
+# export F90=gfortran
```
For clarity please could you add CXX to this list of environment variables.

sergisiso replied:
Done, also added CFLAGS

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step.cpp:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    // Update infrastructure (whole domain)
```
Suggest, "Copy 'next' fields to 'current' fields" rather than "Update infrastructure"

sergisiso replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step.cpp:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    // Time update kernel (internal domain ?)
```
s/?/, u points/

sergisiso replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step.cpp:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    // Time update kernel (internal domain ?)
```
s/?/, v points/

sergisiso replied:
Done

sergisiso commented on benchmarks/nemo/nemolite2d/kernels/c_family/boundary_conditions_kern.c:
```diff
@@ -311,13 +311,14 @@ __kernel void bc_flather_u_code(int width,
 				__global double* restrict sshn_u,
-				__global int* restrict tmask){
+				__global int* restrict tmask,
+                double g){
```
It looked fine for me but I noticed some have tabs and other have spaces, I made the indentation with spaces in the whole file now.

sergisiso commented on benchmarks/nemo/nemolite2d/kernels/c_family/boundary_conditions_kern.c:
```diff
@@ -370,14 +371,15 @@ __kernel void bc_flather_v_code(int width,
 				__global double* restrict sshn_v, 
-				__global int* restrict tmask){
+				__global int* restrict tmask,
+                double g){
```
Same as previous answer, it should be fixed now.

sergisiso commented on benchmarks/nemo/nemolite2d/kernels/c_family/continuity_kern.c:
```diff
@@ -123,13 +124,10 @@ void continuity_code(int ji, int jj,
-    int idx = jj*width + ji;
-
-    if(jj == 0)return;
 
```
I removed it because I didn't know what it does and there isn't any equivalent expression in the Fortran kernel. However, I see there are equivalent expressions in some of the OpenCL pre-processor blocks in other kernels, so I re-introduced it just inside the OpenCL for consistency.
(At some point I will try to use again the 'single_file_c_opencl' version that uses this kernels and see if everything still works and if this expressions can be removed)

sergisiso commented on benchmarks/nemo/nemolite2d/kernels/c_family/momentum_u_kern.c:
```diff
@@ -171,14 +170,14 @@ __kernel void momentum_u_code(int width,
 		     double *e1u, double *e1v, double *e1t,
 		     double *e2u, double *e2t, double *e12u, double *gphiu,
-		     double rdt, double cbfr, double visc){
+		     double rdt, double cbfr, double visc, double omega, double d2r, double g){
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/Makefile:
```diff
@@ -0,0 +1,88 @@
+# Makefile for manual OpenMP version of PSyKAl'd NEMOLite2D.
```
Yes

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/Makefile:
```diff
@@ -0,0 +1,88 @@
+EXECS = nemolite2d
+MODULES = time_step_mod.o
+
+# API lib is an archive that must come at the end of the list of objects
```
I replaced the whole Makefile with the one that includes the `../common.mk`, it has a similar message with INF_LIB, is this more appropriate?

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step_mod.f90:
```diff
@@ -0,0 +1,94 @@
+            sshn_v, hu, hv, un, vn, ua, ht, ssha_u, va, ssha_v
+        INTEGER, intent(in) :: istp
+
+        ! TODO: Should this use %get_data() instead?
```
The compiler didn't like it. It says the get_data() can not be passed to an INOUT parameter. It probably can be solved with a temporary reference that then is passed to c wrapper, but since the %data already works I would prefer to think more carefully about this in a future PR.

arporter replied:
Fine. Thanks for trying.

arporter replied:
Please could you create an issue for this and reference it here.

sergisiso commented:
@arporter This is ready for the next review, I addressed the comments and made the following changes:
- I use now a new-style Makefile (that includes ../common.mk)
- I added a time_step_omp.cpp and time_step_ompopt.cpp for C++ OpenMP and C++ OpenMP optimized to have some more performance comparison points.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step_ompopt.cpp:
```diff
@@ -0,0 +1,144 @@
+
+// Optimizations:
+// - Fuse some kernels (x1.1), fusing more kernels didn't make any difference.
+// - Change array copies with swaps is x1.1 faster, but since the fortran->c
```
I'm a bit confused here - the ptr-swapping code is commented out below and therefore this optimisation is not applied. If that is because of the limitations of fortran->c calling then please say so (here and below).

sergisiso replied:
That's right. I commented it instead of deleting it because I would still try to make it work in the future. As suggested I added comments here and below.

arporter commented:
Just a couple of minor points now. Also `make summary_table` doesn't do include the two new OpenMP versions that you've added. Should it?

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/Makefile:
```diff
@@ -1,76 +1,61 @@
+		$(LDFLAGS) ${OMPFLAGS} -lstdc++
 
+# C++ rules
 time_step.o: time_step.cpp
```
I wasn't sure if this would work as we already have a `%.o: %.f90` but seems Makefile is smart enough to pick the right one. I added a comment warning of it.

sergisiso commented:
@arporter Addressed the comments, ready for next review.

> Also make summary_table doesn't do include the two new OpenMP versions that you've added. Should it?

Maybe it should, the problem is the way I implemented it the first column is printed by the outer Makefile, so the inner Makefile can only print one line respecting the format. I should probably make the inner makefile print whole lines and more than one if it has to. This can be useful for the different fortran serial versions or implementations that can target more than one platform. But since I have to adjust all Makefiles I will leave this for another PR.

arporter commented:
All good now. Thanks for updating.
